plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -241,103 +218,104 @@ activateContract contract = do
         lookupContract @t contract
     activeContractInstanceId <- ContractInstanceId <$> uuidNextRandom
     logInfo . render $ "Initializing contract instance with ID" <+> pretty activeContractInstanceId
-    let currentIteration = iterationZero
+    let initialIteration = succ mempty -- FIXME get max it. from initial response
     response <- Contract.invokeContract @t $ InitContract contractDef
-    sendContractStateMessages @t contract activeContractInstanceId currentIteration response
+    sendContractStateMessages @t contract activeContractInstanceId initialIteration response
     logInfo . render $ "Activated contract instance:" <+> pretty activeContractInstanceId
     pure activeContractInstanceId
 
-processOwnPubkeyRequests ::
+extract :: Alternative f => Prism' a b -> a -> f b
+extract p = maybe empty pure . preview p
+
+-- | Request handlers that can choose whether to handle an effect (using
+--   'Alternative'). This is useful if 'req' is a sum type.
+newtype RequestHandler effs req resp = RequestHandler { unRequestHandler :: req -> Eff (NonDet ': effs) resp }
+    deriving stock (Functor)
+    deriving (Profunctor) via (Kleisli (Eff (NonDet ': effs)))
+    deriving (Semigroup, Monoid) via ((Ap ((->) req) (Alt (Eff (NonDet ': effs)) resp)))
+
+respondtoRequests ::
     forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member Log effs
+    )
+    => RequestHandler effs ContractSCBRequest ContractResponse
+    -> Eff effs ()
+respondtoRequests handler = do
+    state <- fmap (hooks . csCurrentState . getLast) <$> runGlobalQuery (Query.contractState @t)
```

Comment:
For my money, I think if you've got a bunch of stuff before a `<$>` in a do block, I think it's usually worth just binding the stuff to the right of the `<$>` and doing the rest in a `let`.